pkg/controller/resourcedistribution/utils.go

```diff
@@ -0,0 +1,276 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	GetConditionID           = 0
+	CreateConditionID        = 1
+	UpdateConditionID        = 2
+	DeleteConditionID        = 3
+	NoConflictConditionID    = 4
+	NoUncreatedNSConditionID = 5
+	NumberOfConditionTypes   = 6
+	ReconcileSuccessfully    = "Successful"
+)
+
+// UnexpectedError is designed to store the information about .status.conditions when error occurs
+type UnexpectedError struct {
+	err         error
+	namespace   string
+	conditionID int
+}
+
+// isInList return true if namespaceName is in namespaceList, else return false
+func isInList(namespaceName string, namespaceList []appsv1alpha1.ResourceDistributionNamespace) bool {
+	for _, namespace := range namespaceList {
+		if namespaceName == namespace.Name {
+			return true
+		}
+	}
+	return false
+}
+
+// matchViaIncludedNamespaces return true if namespace is in targets.IncludedNamespaces
+func matchViaIncludedNamespaces(namespace *corev1.Namespace, distributor *appsv1alpha1.ResourceDistribution) (bool, error) {
+	if isInList(namespace.Name, distributor.Spec.Targets.IncludedNamespaces.List) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// matchViaLabelSelector return true if namespace matches with target.NamespacesLabelSelectors
+func matchViaLabelSelector(namespace *corev1.Namespace, distributor *appsv1alpha1.ResourceDistribution) (bool, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&distributor.Spec.Targets.NamespaceLabelSelector)
+	if err != nil {
+		return false, err
+	}
+	if !selector.Empty() && selector.Matches(labels.Set(namespace.Labels)) {
+		return true, nil
+	}
+	return false, nil
+}
+
+// matchViaTargets check whether Namespace matches ResourceDistribution via labels
+func matchViaTargets(namespace *corev1.Namespace, distributor *appsv1alpha1.ResourceDistribution) (bool, error) {
+	targets := &distributor.Spec.Targets
+	if isInList(namespace.Name, targets.ExcludedNamespaces.List) {
+		return false, nil
+	}
+	if targets.AllNamespaces {
+		return true, nil
+	}
+	if isInList(namespace.Name, targets.IncludedNamespaces.List) {
+		return true, nil
+	}
+	return matchViaLabelSelector(namespace, distributor)
+}
+
+// hashResource hash resource yaml as version using SHA256
+func hashResource(resourceYaml runtime.RawExtension) string {
+	md5Hash := sha256.Sum256(resourceYaml.Raw)
+	return hex.EncodeToString(md5Hash[:])
+}
+
+// prepareNamespaces return:
+// (1) inTargets: a map that contains all target namespace names
+// (2) allNamespaces: all existing namespace names, including both those fetched from cluster and newly-created namespaces
+// (3) uncreatedNamespaces: the namespaces in targets but not existing
+func prepareNamespaces(handlerClient client.Client, distributor *appsv1alpha1.ResourceDistribution) (inTargets map[string]struct{}, allNamespaces, uncreatedNamespaces []string, err error) {
```

Comment:
The logic here is too complicated.
1. How about merge `prepareNamespaces` and `parseNamespaces` into one `getNamespacesForDistributor` function?
2. Change all `map[string]struct{}` to `sets.String`.
3. `inTargets`/`allNamespaces`/`uncreatedNamespaces` are not clear. How about just `matchedNamespaces` and `unmatchedNamespaces`? Then in `doReconcile` there can be two `for` loop, one for `matchedNamespaces` to ensure objects in them and another for `unmatchedNamespaces` to delete objects created in them.

Reply:
ok, I'll try to make this logic more clear.